SmartDeviceLink/SDLMenuManager.m

```diff
@@ -0,0 +1,400 @@
+@property (assign, nonatomic) BOOL waitingOnHMILevelUpdate;
+
+@property (assign, nonatomic) UInt32 lastMenuId;
+@property (copy, nonatomic) NSArray<SDLMenuCell *> *oldMenuCells;
+
+@end
+
+UInt32 const ParentIdNotFound = UINT32_MAX;
+UInt32 const MenuCellIdMin = 1;
+
+@implementation SDLMenuManager
+
+- (instancetype)init {
+    self = [super init];
+    if (!self) { return nil; }
+
+    _lastMenuId = MenuCellIdMin;
+    _menuCells = @[];
+    _oldMenuCells = @[];
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_registerResponse:) name:SDLDidReceiveRegisterAppInterfaceResponse object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_displayLayoutResponse:) name:SDLDidReceiveSetDisplayLayoutResponse object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_hmiStatusNotification:) name:SDLDidChangeHMIStatusNotification object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_commandNotification:) name:SDLDidReceiveCommandNotification object:nil];
+
+    return self;
+}
+
+- (instancetype)initWithConnectionManager:(id<SDLConnectionManagerType>)connectionManager fileManager:(SDLFileManager *)fileManager {
+    self = [self init];
+    if (!self) { return nil; }
+
+    _connectionManager = connectionManager;
+    _fileManager = fileManager;
+
+    return self;
+}
+
+#pragma mark - Updating System
+
+- (void)sdl_updateWithCompletionHandler:(nullable SDLMenuUpdateCompletionHandler)completionHandler {
+    if (self.currentLevel == nil || [self.currentLevel isEqualToString:SDLHMILevelNone]) {
+        return;
+    }
+
+    if (self.inProgressUpdate != nil) {
+        // There's an in progress update, we need to put this on hold
+        self.hasQueuedUpdate = YES;
+        return;
+    }
+
+    __weak typeof(self) weakself = self;
+    [self sdl_sendDeleteCurrentMenu:^(NSError * _Nullable error) {
+        [weakself sdl_sendCurrentMenu:^(NSError * _Nullable error) {
+            weakself.inProgressUpdate = nil;
+
+            if (completionHandler != nil) {
+                completionHandler(error);
+            }
+
+            if (weakself.hasQueuedUpdate) {
+                [weakself sdl_updateWithCompletionHandler:nil];
+                weakself.hasQueuedUpdate = NO;
+            }
+        }];
+    }];
+}
+
+#pragma mark Delete Old Menu Items
+
+- (void)sdl_sendDeleteCurrentMenu:(SDLMenuUpdateCompletionHandler)completionHandler {
+    if (self.oldMenuCells.count == 0) {
+        completionHandler(nil);
+
+        return;
+    }
+
+    NSArray<SDLRPCRequest *> *deleteMenuCommands = [self sdl_deleteCommandsForCells:self.oldMenuCells];
+    self.oldMenuCells = @[];
+
+    [self.connectionManager sendRequests:deleteMenuCommands progressHandler:nil completionHandler:^(BOOL success) {
+        if (!success) {
+            SDLLogE(@"Error deleting old menu commands");
```

Comment:
If the `DeleteSubMenu` command fails, the `cellId` should not be reused by the manager. 

If I update the menu while in a sub-menu, I get the following error:

```
{
    info = "Sub Menu is in Use";
    resultCode = "IN_USE";
    success = 0;
}
```
Then when I try to update the menu again, I get a bunch of `INVALID_ID` errors due to the manager trying to reuse the `cellId`.

Reply:
Should be fixed